cpp/include/raft/neighbors/detail/ivf_flat_build.cuh

```diff
@@ -415,4 +417,103 @@ inline void fill_refinement_index(raft::device_resources const& handle,
                                          refinement_index->veclen());
   RAFT_CUDA_TRY(cudaPeekAtLastError());
 }
+
+template <typename T, typename IdxT>
+__global__ void get_data_ptr_kernel(const uint32_t* list_sizes,
+                                    const T* const* list_data_ptrs,
+                                    const IdxT* const* list_indices_ptrs,
+                                    uint32_t dim,
+                                    uint32_t veclen,
+                                    uint32_t n_list,
+                                    IdxT max_indice,
+                                    T** ptrs_to_data)
+{
+  const IdxT list_id = IdxT(blockDim.x) * IdxT(blockIdx.x) + threadIdx.x;
+  if (list_id >= n_list) { return; }
+  const IdxT inlist_id     = IdxT(blockDim.y) * IdxT(blockIdx.y) + threadIdx.y;
+  const uint32_t list_size = list_sizes[list_id];
+  if (inlist_id >= list_size) { return; }
+
+  auto* list_indices = list_indices_ptrs[list_id];
+  IdxT id            = list_indices[inlist_id];
+  if (id > max_indice) { return; }
+
+  using interleaved_group = Pow2<kIndexGroupSize>;
+  auto group_offset       = interleaved_group::roundDown(inlist_id);
+  auto ingroup_id         = interleaved_group::mod(inlist_id) * veclen;
+
+  auto* list_data  = list_data_ptrs[list_id];
+  const T* ptr     = list_data + (group_offset * dim) + ingroup_id;
+  ptrs_to_data[id] = (T*)ptr;
+}
+
+template <typename T, typename IdxT>
+__global__ void reconstruct_batch_kernel(const IdxT* vector_ids,
+                                         const T** ptrs_to_data,
+                                         uint32_t dim,
+                                         uint32_t veclen,
+                                         IdxT n_rows,
+                                         T* reconstr)
+{
+  const IdxT i = IdxT(blockDim.x) * IdxT(blockIdx.x) + threadIdx.x;
+  if (i >= n_rows) { return; }
+
+  const IdxT vector_id = vector_ids[i];
+  const T* src         = ptrs_to_data[vector_id];
+  if (!src) { return; }
+
+  reconstr += i * dim;
+  for (uint32_t l = 0; l < dim; l += veclen) {
+    for (uint32_t j = 0; j < veclen; j++) {
+      reconstr[l + j] = src[l * kIndexGroupSize + j];
+    }
+  }
+}
+
+template <typename T, typename IdxT>
+void reconstruct_batch(raft::device_resources const& handle,
+                       const index<T, IdxT>& index,
+                       device_mdspan<const IdxT, extent_1d<IdxT>, row_major> vector_ids,
+                       device_mdspan<T, extent_2d<IdxT>, row_major> vector_out)
+{
+  thrust::device_ptr<const IdxT> vector_ids_ptr =
+    thrust::device_pointer_cast(vector_ids.data_handle());
+  IdxT max_indice = *thrust::max_element(
```

Comment:
Just a reminder: although seeming like a good compromise, this would be wasting too much memory at best and crash at worst when the database contains non-contiguous (e.g. beyond a few billions) 64 bit user-space indices.
We've discussed this in the corresponding ivf-pq PR https://github.com/rapidsai/raft/pull/1298#discussion_r1127379995 and I suggested to create a special mapping struct (a hashmap or a pair of sorted arrays) for faster search of the user-space indices in https://github.com/rapidsai/raft/issues/1205 .

I think, we'd better reconstruct by cluster+offset pairs here and dedicate a separate PR for the mapping between user-space and DB indices, and put it into `ivf_list.cuh` to share among the ivf implementations.